src/dms/core/EdFi.DataManagementService.Core/OpenApi/OpenApiDocument.cs

```diff
@@ -157,79 +157,1068 @@ private void InsertExts(
+    }
 
-            JsonObject extensionProperties =
-                extensionSchema["properties"]?.AsObject()
-                ?? throw new InvalidOperationException(
-                    $"Extension schema '{extensionSchemaName}' missing 'properties'."
+    /// <summary>
+    /// Sets up the extension schema structure
+    /// </summary>
+    private static void SetupExtensionSchema(
+        JsonObject schemaLocation,
+        JsonObject componentsSchemas,
+        ExtensionSchemaNames schemaNames,
+        string projectName
+    )
+    {
+        AddExtensionReference(schemaLocation, schemaNames.ExtensionSchemaName);
+        CreateExtensionSchemaIfNeeded(componentsSchemas, schemaNames.ExtensionSchemaName);
+        AddProjectReference(componentsSchemas, schemaNames, projectName);
+    }
+
+    /// <summary>
+    /// Adds _ext reference to main schema
+    /// </summary>
+    private static void AddExtensionReference(JsonObject schemaLocation, string extensionSchemaName)
+    {
+        if (schemaLocation["_ext"] == null)
+        {
+            schemaLocation.Add(
+                "_ext",
+                JsonNode.Parse($"{{ \"$ref\": \"#/components/schemas/{extensionSchemaName}\" }}")
+            );
+        }
+    }
+
+    /// <summary>
+    /// Creates extension schema if it doesn't exist
+    /// </summary>
+    private static void CreateExtensionSchemaIfNeeded(
+        JsonObject componentsSchemas,
+        string extensionSchemaName
+    )
+    {
+        if (!componentsSchemas.ContainsKey(extensionSchemaName))
+        {
+            componentsSchemas.Add(
+                extensionSchemaName,
+                new JsonObject { ["type"] = "object", ["properties"] = new JsonObject() }
+            );
+        }
+    }
+
+    /// <summary>
+    /// Adds project reference to extension schema
+    /// </summary>
+    private static void AddProjectReference(
+        JsonObject componentsSchemas,
+        ExtensionSchemaNames schemaNames,
+        string projectName
+    )
+    {
+        var extensionSchema = componentsSchemas[schemaNames.ExtensionSchemaName]!.AsObject();
+        var extensionProperties = extensionSchema["properties"]!.AsObject();
+
+        if (!extensionProperties.ContainsKey(projectName))
+        {
+            extensionProperties.Add(
+                projectName,
+                JsonNode.Parse(
+                    $"{{ \"$ref\": \"#/components/schemas/{schemaNames.ProjectExtensionSchemaName}\" }}"
+                )
+            );
+        }
+    }
+
+    /// <summary>
+    /// Adds project-specific schema to components
+    /// </summary>
+    private static void AddProjectSchema(JsonObject componentsSchemas, string schemaName, JsonObject schema)
+    {
+        if (!componentsSchemas.ContainsKey(schemaName))
+        {
+            componentsSchemas.Add(schemaName, schema);
+        }
+    }
+
+    // Supporting types and records
+    private sealed record ExtensionContext(
```

Comment:
Split all these object definitions into separate files and provide documentation on what they are